### PR TITLE
VM: implement stack item conversion

### LINKED ISSUE #372: VM: implicit operand conversion
In the reference NeoVM implementation arithmetic opcodes (`SIGN`, `ABS` ...) can also work with boolean values.
https://github.com/neo-project/neo-vm/blob/master/tests/neo-vm.Tests/Tests/OpCodes/Numeric/SIGN.json#L31
This is because `GetBigInteger` is defined for `Boolean` stack item.

Here is a list of implicit conversions than can be performed:
1. `Integer` from `Integer` and `Boolean` . (and default using `ByteArray`)
2. `ByteArray` from `Integer`, `ByteArray` and `Boolean`
3. `Boolean` from `ByteArray`, `Integer`, `Boolean`, `Array`, `Interop` and `Map`

Links:
Interface https://github.com/neo-project/neo-vm/blob/f9a5b2fa0f806b7f9db542efea4fcce9aae917b2/src/neo-vm/StackItem.cs
Implementations https://github.com/neo-project/neo-vm/tree/bf891cf07f7a2cd25212050cc6fe20588f77934d/src/neo-vm/Types

### PR DESCRIPTION
**Pull request by fyrchik**
Fixes #372.

Also added tests for `NOT` and `SIGN` to check new behavior.
Also fixed a bug with `Bool` casting: it should be `!= 0` instead of `== 1`.

There are still missing cases for `Interop` and `Map` because they are not implemented.

### CODE REVIEW

File: pkg/vm/stack.go [hunk 1]
```diff
@@ -57,6 +57,24 @@ func (e *Element) Prev() *Element {
 	return nil
 }
 
+// ByteArray attempts to get the underlying value of the element as a bytearray.
+// Will panic if the assertion failed which will be caught by the VM.
+func (e *Element) ByteArray() []byte {
```
Comment (roman-khimov) [inline pkg/vm/stack.go]: Are there any users for it? It's strange to see it added without any real use.

**Comment by roman-khimov** [comment]
`vm: implement BigInt -> Boolean conversion ` in 52b0d661e065e52380c797ebef9fc9e1da1ae11e, I guess it should be the other way around: `vm: implement Boolean -> BigInt conversion `

File: pkg/vm/stack.go [hunk 2]
```diff
@@ -95,10 +95,23 @@ func (e *Element) BigInt() *big.Int {
...
+		return true
+	case *ByteArrayItem:
+		for _, b := range t.value {
+			if b != 0 {
+				return true
+			}
+		}
+		return false
+	default:
+		panic("unsupported")
```
Comment (roman-khimov) [inline pkg/vm/stack.go]: It would be nice to say something about operand type here, plain `unsupported` is not informative and user-friendly.

File: pkg/vm/stack.go [hunk 3]
```diff
@@ -57,6 +57,24 @@ func (e *Element) Prev() *Element {
...
+		return t.value
+	case *BigIntegerItem:
+		return util.ArrayReverse(t.value.Bytes()) // neoVM returns in LE
+	case *BoolItem:
+		if t.value {
+			return []byte{1}
+		}
+		return []byte{0}
+	default:
+		panic("unsupported")
```
Comment (roman-khimov) [inline pkg/vm/stack.go]: Same thing here.

File: pkg/vm/stack.go (hunk 2 above)
Comment (fyrchik) [reply pkg/vm/stack.go]: fixed

File: pkg/vm/stack.go (hunk 3 above)
Comment (fyrchik) [reply pkg/vm/stack.go]: fixed

File: pkg/vm/stack.go (hunk 1 above)
Comment (fyrchik) [reply pkg/vm/stack.go]: Removed it. I forgot about already existing `Bytes()`.
In С# implementation `CAT` opcode can involve conversions to `ByteArray`

**Comment by roman-khimov** [comment]
OK, so now please split  0a78490c822e8889f3a4a203394a836db9aec8a4 into two parts --- the one about `Bytes()` (that should be squashed with 36b7d37128123004d2a633f2cab8efd75b3f0721) and the one about CAT tests and then squash 246ea95a331daa64d006b1c733b7b483b1a4f835 with 78c32393120bac44d5bd89768cc5c7f94183bbf5.

File: pkg/vm/stack.go [hunk 4]
```diff
@@ -92,7 +92,7 @@ func (e *Element) Bool() bool {
 		}
 		return false
 	default:
-		panic("unsupported")
+		panic("can't convert to bool: " + t.String())
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/vm/stack.go]: Issues of this PR created by other participants :)
The whole commit history is supposed to be cleaned before merge.

File: pkg/vm/vm_test.go [hunk 5]
```diff
@@ -729,6 +729,28 @@ func TestCATGood(t *testing.T) {
 	assert.Equal(t, []byte("abcdef"), vm.estack.Peek(0).Bytes())
 }
 
+func TestCATInt0Bytearray(t *testing.T) {
```
Comment (volekerb) [inline pkg/vm/vm_test.go]: TestCATInt0Bytearray and TestCATBytearrayInt1: Suggest to use same naming convention and have camel case like:
TestCATInt0ByteArray and TestCATInt1ByteArray
or
TestCATByteArrayInt0 and TestCATByteArrayInt1

File: pkg/vm/stack.go [hunk 6]
```diff
@@ -117,7 +99,19 @@ func (e *Element) Bool() bool {
 // Bytes attempts to get the underlying value of the element as a byte array.
 // Will panic if the assertion failed which will be caught by the VM.
 func (e *Element) Bytes() []byte {
-	return e.value.Value().([]byte)
+	switch t := e.value.(type) {
```
Comment (volekerb) [inline pkg/vm/stack.go]: I believe these changes  should be in separate commit not in "add tests for CAT"

File: pkg/vm/vm_test.go (hunk 5 above)
Comment (fyrchik) [reply pkg/vm/vm_test.go]: The order in name mirrors the order of arguments on stack.

File: pkg/vm/vm_test.go [hunk 7]
```diff
@@ -111,6 +111,67 @@ func TestPushData4(t *testing.T) {
...
+func TestNOTStruct(t *testing.T) {
+	prog := makeProgram(NOT)
+	vm := load(prog)
+	vm.estack.Push(NewElement(&StructItem{[]StackItem{}}))
+	vm.Run()
+	assert.Equal(t, false, vm.state.HasFlag(faultState))
+	assert.Equal(t, &BoolItem{false}, vm.estack.Pop().value)
+}
+
+func TestNOTByteArray0(t *testing.T) {
```
Comment (volekerb) [inline pkg/vm/vm_test.go]: just curios: Is there any way we can group test in order not to duplicate code.
Having smth like:
```
tests := []struct {
		name    string
		args    args
		want    []byte
		wantErr bool
	}
t.Run(tt.name, func(t *testing.T) {..}
```
What do you think?

File: pkg/vm/vm_test.go (hunk 5 above)
Comment (roman-khimov) [reply pkg/vm/vm_test.go]: But `ByteArray` is still supposed to be `ByteArray`, I think.

File: pkg/vm/vm_test.go (hunk 7 above)
Comment (roman-khimov) [reply pkg/vm/vm_test.go]: Maybe. Or maybe it all is gonna be irrelevant when we're to complete #196.
